File: week02/example/ltc78_subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        subset = []
        res = []

        def recur(nums, i):
            if i == n:
                res.append(subset.copy())
                return

            recur(nums, i + 1)
            subset.append(nums[i])
            recur(nums, i + 1)
            subset.pop()

        recur(nums, 0)
        return res


if __name__ == '__main__':
    print(Solution().subsets([1, 2, 3]))

File: week02/example/ltc78_subsets.cpp
//
// Created by chenbingwei on 2022/4/29.
//

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int> &nums) {
        recur(nums, 0);
        return ans;
    }

private:
    vector<vector<int>> ans;
    vector<int> subset;

    void recur(vector<int> &nums, int i) {
        if (i == nums.size()) {
            ans.push_back(subset);
            return;
        }

        recur(nums, i + 1);
        subset.push_back(nums[i]);
        recur(nums, i + 1);
        subset.pop_back();

    }
};
